### PR TITLE
Feature/guava programming

### PR DESCRIPTION
**Pull request by sauravbiswasiupr**
- Implemented basic string split, join utilities
- Implemented basic functional programming utilities to read data from hashmaps.
- Tests 🍻

### CODE REVIEW

File: src/test/java/GuavaProgramming/StringProcessingUtilityTest.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+    @Test
+    public void joinStringListTest() throws Exception {
+        List<String> strings = Lists.newArrayList("I", "am", "good");
+        Assert.assertTrue(sp.joinStringList(strings).equals("I,am,good"));
+    }
+
+    @Test
+    public void joinMapAsStringTest() {
+        Map<String, String> testMap = Maps.newLinkedHashMap();
```
Comment (sougatabhattacharjee) [inline src/test/java/GuavaProgramming/StringProcessingUtilityTest.java]: final

File: src/test/java/GuavaProgramming/StringProcessingUtilityTest.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        testMap.put("messi", "barcelona");
+        testMap.put("ronaldo", "real");
+        testMap.put("griezmann", "atletico");
+
+        Assert.assertEquals(sp.joinMapAsString(testMap), "messi=barcelona#ronaldo=real#griezmann=atletico");
+    }
+
+    @Test
+    public void splitStringAsListTest() throws NullPointerException {
+        String toSplit="foo|bar|baz ";
```
Comment (sougatabhattacharjee) [inline src/test/java/GuavaProgramming/StringProcessingUtilityTest.java]: final

File: src/test/java/GuavaProgramming/StringProcessingUtilityTest.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        sp = new StringProcessingUtility();
+    }
+
+    @After
+    public void tearDown() throws Exception {
+    }
+
+    @Test
+    public void joinStringListTest() throws Exception {
+        List<String> strings = Lists.newArrayList("I", "am", "good");
```
Comment (sougatabhattacharjee) [inline src/test/java/GuavaProgramming/StringProcessingUtilityTest.java]: final

File: src/test/java/GuavaProgramming/StringProcessingUtilityTest.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import java.util.Map;
+
+/**
+ * Created by saurav on 9/7/16.
+ */
+public class StringProcessingUtilityTest {
+    StringProcessingUtility sp;
+    @Before
+    public void setUp() throws Exception {
+        sp = new StringProcessingUtility();
```
Comment (sougatabhattacharjee) [inline src/test/java/GuavaProgramming/StringProcessingUtilityTest.java]: Actually u can skip this and initialize the object beginning. 
final StringProcessingUtility sp = new StringProcessingUtility();

File: src/test/java/GuavaProgramming/FunctionalUtilitiesTest.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    @After
+    public void tearDown() throws Exception {
+
+    }
+
+    @Test
+    public void returnKeyValueTest() throws Exception {
+        Map<String, List> dataMap = Maps.newHashMap();
```
Comment (sougatabhattacharjee) [inline src/test/java/GuavaProgramming/FunctionalUtilitiesTest.java:33]: May be better to use generics here.
Map<String, List<String>> dataMap = Maps.newHashMap();

File: src/test/java/GuavaProgramming/FunctionalUtilitiesTest.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        dataMap.put("messi", Lists.newArrayList("barcelona", "argentina", "newell's"));
+        dataMap.put("ronaldo", Lists.newArrayList("real madrid", "portugal", "sporting"));
+        funcUtils.setDataMap(dataMap);
+
+        String expectedMessi = "barcelona,argentina,newell's";
```
Comment (sougatabhattacharjee) [inline src/test/java/GuavaProgramming/FunctionalUtilitiesTest.java]: final

File: src/test/java/GuavaProgramming/FunctionalUtilitiesTest.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        String expectedRonaldo = "real madrid,portugal,sporting";
```
Comment (sougatabhattacharjee) [inline src/test/java/GuavaProgramming/FunctionalUtilitiesTest.java]: final

File: src/test/java/GuavaProgramming/FunctionalUtilitiesTest.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+        Assert.assertEquals(expectedMessi, funcUtils.returnKeyValue("messi"));
```
Comment (sougatabhattacharjee) [inline src/test/java/GuavaProgramming/FunctionalUtilitiesTest.java:41]: consider using static import, which makes things more meaningful

File: src/main/java/GuavaProgramming/FunctionalUtilities.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+    private final class KeyValuesListToString implements Function<String, String> {
+        public String apply(String key) {
+            List<String> values = dataMap.get(key);
+            return joiner.skipNulls().join(values);
+        }
+    }
+
+    public String returnKeyValue(String key) {
+        Preconditions.checkArgument((dataMap.size() > 0), "datamap has to be initialized");
```
Comment (sougatabhattacharjee) [inline src/main/java/GuavaProgramming/FunctionalUtilities.java]: Not sure about this checks applicable here.

File: src/test/java/CommonAlgorithms/MaxSumIncreasingSubsequenceTest.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    @After
+    public void tearDown() throws Exception {
+
+    }
+
+    @Test
+    public void maxSumSubSeqLength() throws Exception {
+        int[] arr = {1, 101, 2, 3, 100, 4, 5};
```
Comment (sougatabhattacharjee) [inline src/test/java/CommonAlgorithms/MaxSumIncreasingSubsequenceTest.java]: final

File: src/test/java/GuavaProgramming/PredicateUtilityTest.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.List;
+
+import static org.junit.Assert.*;
```
Comment (sougatabhattacharjee) [inline src/test/java/GuavaProgramming/PredicateUtilityTest.java]: dont use general imports (*), use specific imports. 
If you are using intellij please consider following settings.
intellij autoimport

File: src/test/java/GuavaProgramming/PredicateUtilityTest.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    @After
+    public void tearDown() throws Exception {
+
+    }
+
+    @Test
+    public void testPredicateUtility() {
+        List<String> strings = Lists.newArrayList("saurav", "sougata", "deep", "debopriyo");
```
Comment (sougatabhattacharjee) [inline src/test/java/GuavaProgramming/PredicateUtilityTest.java]: final

File: src/test/java/GuavaProgramming/PredicateUtilityTest.java [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+        Iterable<String> filtered = Iterables.filter(strings, predicateUtility);
```
Comment (sougatabhattacharjee) [inline src/test/java/GuavaProgramming/PredicateUtilityTest.java]: final

File: src/test/java/CommonAlgorithms/MaxSumIncreasingSubsequenceTest.java [hunk 14]
```diff
@@ -0,0 +1,31 @@
+package CommonAlgorithms;
+
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.junit.Assert.*;
```
Comment (sougatabhattacharjee) [inline src/test/java/CommonAlgorithms/MaxSumIncreasingSubsequenceTest.java]: See the other comment

File: src/test/java/GuavaProgramming/FunctionalUtilitiesTest.java (hunk 5 above)
Comment (sauravbiswasiupr) [reply src/test/java/GuavaProgramming/FunctionalUtilitiesTest.java:33]: no it is actually the behavior i want, the String utility must work on a map that has strings and gives me lists back.

File: src/test/java/GuavaProgramming/FunctionalUtilitiesTest.java (hunk 8 above)
Comment (sauravbiswasiupr) [reply src/test/java/GuavaProgramming/FunctionalUtilitiesTest.java:41]: Okay, but seems like overkill. Can you be a little more explicit ?

File: src/test/java/CommonAlgorithms/MaxSumIncreasingSubsequenceTest.java (hunk 14 above)
Comment (sauravbiswasiupr) [reply src/test/java/CommonAlgorithms/MaxSumIncreasingSubsequenceTest.java]: ✅

File: src/test/java/GuavaProgramming/PredicateUtilityTest.java (hunk 13 above)
Comment (sauravbiswasiupr) [reply src/test/java/GuavaProgramming/PredicateUtilityTest.java]: ✅

File: src/test/java/GuavaProgramming/PredicateUtilityTest.java (hunk 12 above)
Comment (sauravbiswasiupr) [reply src/test/java/GuavaProgramming/PredicateUtilityTest.java]: ✅

File: src/test/java/GuavaProgramming/PredicateUtilityTest.java (hunk 11 above)
Comment (sauravbiswasiupr) [reply src/test/java/GuavaProgramming/PredicateUtilityTest.java]: ✅

File: src/test/java/CommonAlgorithms/MaxSumIncreasingSubsequenceTest.java (hunk 10 above)
Comment (sauravbiswasiupr) [reply src/test/java/CommonAlgorithms/MaxSumIncreasingSubsequenceTest.java]: ✅

File: src/main/java/GuavaProgramming/FunctionalUtilities.java (hunk 9 above)
Comment (sauravbiswasiupr) [reply src/main/java/GuavaProgramming/FunctionalUtilities.java]: removed ✅

File: src/test/java/GuavaProgramming/FunctionalUtilitiesTest.java (hunk 7 above)
Comment (sauravbiswasiupr) [reply src/test/java/GuavaProgramming/FunctionalUtilitiesTest.java]: ✅

File: src/test/java/GuavaProgramming/FunctionalUtilitiesTest.java (hunk 6 above)
Comment (sauravbiswasiupr) [reply src/test/java/GuavaProgramming/FunctionalUtilitiesTest.java]: ✅

File: src/test/java/GuavaProgramming/StringProcessingUtilityTest.java (hunk 4 above)
Comment (sauravbiswasiupr) [reply src/test/java/GuavaProgramming/StringProcessingUtilityTest.java]: ✅

File: src/test/java/GuavaProgramming/StringProcessingUtilityTest.java (hunk 3 above)
Comment (sauravbiswasiupr) [reply src/test/java/GuavaProgramming/StringProcessingUtilityTest.java]: ✅

File: src/test/java/GuavaProgramming/StringProcessingUtilityTest.java (hunk 2 above)
Comment (sauravbiswasiupr) [reply src/test/java/GuavaProgramming/StringProcessingUtilityTest.java]: ✅

File: src/test/java/GuavaProgramming/StringProcessingUtilityTest.java (hunk 1 above)
Comment (sauravbiswasiupr) [reply src/test/java/GuavaProgramming/StringProcessingUtilityTest.java]: ✅

**Pull request merged**